pytorch_lightning/trainer/connectors/data_connector.py

```diff
@@ -117,9 +117,12 @@ def attach_datamodule(self, model, datamodule, stage):
             if is_overridden('test_dataloader', datamodule):
                 model.test_dataloader = datamodule.test_dataloader
 
-            # Override transfer_batch_to_device if dataset-specific to_device logic has been defined in datamodule
-            if is_overridden('transfer_batch_to_device', datamodule):
-                model.transfer_batch_to_device = datamodule.transfer_batch_to_device
+            # TODO check this before merge
+            # Override prepare_batch_for_transfer if dataset-specific to_device logic has been defined in datamodule
+            model.prepare_batch_for_transfer = datamodule.prepare_batch_for_transfer
+
+            # TODO remove this after all the changes are done
+            model.transfer_batch_to_device = datamodule.transfer_batch_to_device
```

Comment:
I understand what you try to achieve here. It is done in other places. 

However, I am not a fan of patching the model methods through the Trainer. 
This is already done with other methods, like the dataloaders :(
The bad thing I see here is that if I have a method `module.method`,
during `trainer.fit(model)` Trainer replaces `module.method = method2` without me knowing, 
now I expect to be able to run `model.method` after fit but it behaves differently. Why? because my model DEFINITION was altered behind the scenes.
In my view, this is currently an issue in Lightning, and I will bring it up again after v1.0. I would very much prefer not to do this patching. After all, this LightningDataModule is supposed to be separate from LightningModule, that's why it was made. The trainer should USE the methods in it for training, but NOT substitute methods in my LightningModule. I hope you can follow my thoughts.

Reply:
yep totally agreed, for now LightningModule is supposed to have some methods of LightningDataModule and it might cause conflict here if someone want's to override the method. That's why marked it as TODO for now. Will revert these changes before merge.